### PR TITLE
[TASK] Clarify variable scope withing TCA files (#4754)

### PR DESCRIPTION
**Pull request by dogawaf**
References TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/issues/4756
Releases: main,12.4

I don't know if a backport to older version should be done.
As a developper, I would also like to clarify the behavior for 11.5 LTS and maybe 10.4. Tell me if and how to proceed.

### CODE REVIEW

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst [hunk 1]
```diff
@@ -14,6 +14,14 @@ The folder :file:`EXT:my_extension/Configuration/TCA/` may contain or override
 All files in this directory are automatically included during the TYPO3
 :ref:`bootstrap <bootstrapping>`.
 
+..  versionadded:: 12.0
+    Files within :file:`Configuration/TCA/` files are loaded within a dedicated scope.
+    This means that variables defined in those files can't leak into the following files.
+
+    ..  note::
+        Until TYPO3 v12, variables declared in these files were in a shared scope,
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:22]: "Until TYPO3 v12" can be misunderstood. At first read, I assumed, in TYPO3 v12 the content has to be wrapped (when using local variables). But this seems not the case. So I suggest something like: "Until TYPO3 v11 and below" ..." to make it clearer.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst [hunk 2]
```diff
@@ -14,6 +14,14 @@ The folder :file:`EXT:my_extension/Configuration/TCA/` may contain or override
 All files in this directory are automatically included during the TYPO3
 :ref:`bootstrap <bootstrapping>`.
 
+..  versionadded:: 12.0
+    Files within :file:`Configuration/TCA/` files are loaded within a dedicated scope.
+    This means that variables defined in those files can't leak into the following files.
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:19]: Does an average user understand this? "leak into the following files". What does that mean? A user may only have one TCA/Overrides files, so there may be no following files.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst [hunk 3]
```diff
@@ -14,6 +14,14 @@ The folder :file:`EXT:my_extension/Configuration/TCA/` may contain or override
 All files in this directory are automatically included during the TYPO3
 :ref:`bootstrap <bootstrapping>`.
 
+..  versionadded:: 12.0
+    Files within :file:`Configuration/TCA/` files are loaded within a dedicated scope.
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:18]:
```suggestion
    Files within :file:`Configuration/TCA/Overrides/` are loaded within a dedicated scope within TYPO3.
```

File: Makefile [hunk 4]
```diff
@@ -9,6 +9,15 @@ docs: ## Generate projects documentation (from "Documentation" directory)
 
 	docker run --rm --pull always -v "$(shell pwd)":/project -t ghcr.io/typo3-documentation/render-guides:latest --config=Documentation
 
+.PHONY: docs-live
```
Comment (brotkrueml) [inline Makefile:12]: This change does not belong to this PR.

**Review by brotkrueml — CHANGES_REQUESTED** [review]
Thanks for your patch. However, I have some remarks.

Please also give a hint when this change was done (code or patchset), or is it verified somewhere?

**Comment by dogawaf** [comment]
> Please also give a hint when this change was done (code or patchset), or is it verified somewhere?

Here is the associated change: https://review.typo3.org/c/Packages/TYPO3.CMS/+/73731
It was merged on v12.0.0

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst (hunk 3 above)
Comment (dogawaf) [reply Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:18]: The whole TCA directory (with Overrides subdirectory) is targetted by the described behavior.
Should I mention both directories?

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst (hunk 1 above)
Comment (dogawaf) [reply Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:22]: Ack

File: Makefile (hunk 4 above)
Comment (dogawaf) [reply Makefile:12]: Oops.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:19]: Sounds better for me.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst (hunk 3 above)
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:18]: In the TCA directory (without Overrides) you are returning an array, so there should be no leakage at all. But maybe someone defines variables there using then in the TCA definition. But that would be an edge case. So, would be okay. Thanks.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst [hunk 5]
```diff
@@ -14,6 +14,14 @@ The folder :file:`EXT:my_extension/Configuration/TCA/` may contain or override
...
+    This means that variables defined in those files can't leak into the following files.
+
+    ..  note::
+        In TYPO3 v11 and below, variables declared in these files were in a shared scope,
+        with the risk of a leakage to the following files. The use of :php:`call_user_func`
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:23]:
```suggestion
        with the risk of a leakage to the following files. The use of :php:`call_user_func()`
```

**Comment by brotkrueml** [comment]
> Here is the associated change: https://review.typo3.org/c/Packages/TYPO3.CMS/+/73731 It was merged on v12.0.0

Thanks.

File: Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst (hunk 2 above)
Comment (linawolf) [inline Documentation/ExtensionArchitecture/FileStructure/Configuration/TCA/Index.rst:19]:
```suggestion
    This means that variables defined in those files can't leak to any other 
    TCA file during the TCA compilation process.
```

**Review by linawolf — APPROVED** [review]

**Review by linawolf — APPROVED** [review]

**Comment by linawolf** [comment]
Thanks a lot for clearing this up! It seems like there was never a changelog-entry attached to this, so the docs team missed the change. 

Did you search if there are other places where we still do this?

**Comment by dogawaf** [comment]
> Thanks a lot for clearing this up! It seems like there was never a changelog-entry attached to this, so the docs team missed the change.
> 
> Did you search if there are other places where we still do this?

@linawolf , I see 3 unneeded `call_user_func` calls there: https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/ContentElements/AddingYourOwnContentElements.html

I will push a new pull request for that.

Do you think we need to specify also there that "since TYPO3 12 blah blah variable scope blah blah"?
